### PR TITLE
Fix unsafe list manipulation in BeatmapDifficultyCache

### LINKED ISSUE #10698: Fatal crash in BeatmapDifficultyManager
```[runtime] 2020-11-05 09:26:24 [error]: System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
[runtime] 2020-11-05 09:26:24 [error]: at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
[runtime] 2020-11-05 09:26:24 [error]: at System.Collections.Generic.List`1.Enumerator.MoveNext()
[runtime] 2020-11-05 09:26:24 [error]: at osu.Game.Beatmaps.BeatmapDifficultyManager.cancelTrackedBindableUpdate() in /Users/dean/Projects/osu/osu.Game/Beatmaps/BeatmapDifficultyManager.cs:line 179
[runtime] 2020-11-05 09:26:24 [error]: at osu.Game.Beatmaps.BeatmapDifficultyManager.updateTrackedBindables() in /Users/dean/Projects/osu/osu.Game/Beatmaps/BeatmapDifficultyManager.cs:line 157
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at System.Collections.Generic.List`1.Enumerator.MoveNext()
   at osu.Game.Beatmaps.BeatmapDifficultyManager.cancelTrackedBindableUpdate() in /Users/dean/Projects/osu/osu.Game/Beatmaps/BeatmapDifficultyManager.cs:line 179
   at osu.Game.Beatmaps.BeatmapDifficultyManager.updateTrackedBindables() in /Users/dean/Projects/osu/osu.Game/Beatmaps/BeatmapDifficultyManager.cs:line 157
   at osu.Game.Beatmaps.BeatmapDifficultyManager.<LoadComplete>b__15_0(ValueChangedEvent`1 _) in /Users/dean/Projects/osu/osu.Game/Beatmaps/BeatmapDifficultyManager.cs:line 50```

### PR DESCRIPTION
**Pull request by smoogipoo**
Should fix https://github.com/ppy/osu/issues/10698

### CODE REVIEW

**Comment by bdach** [comment]
If this doesn't fix it I don't know what will.

That said I'm not 100% sure on the lock in `GetBindableDifficulty()` - one of the usages of it should be fine (`DifficultyIcon.DifficultyRetriever`, inside of `load()`) but the others (beatmap wedge and `DrawableCarouselBeatmap`) might be less so with high contention. Not sure if worth worrying about, though.

File: osu.Game/Beatmaps/BeatmapDifficultyCache.cs [hunk 1]
```diff
@@ -31,7 +31,7 @@ public class BeatmapDifficultyCache : MemoryCachingComponent<BeatmapDifficultyCa
         private readonly ThreadedTaskScheduler updateScheduler = new ThreadedTaskScheduler(1, nameof(BeatmapDifficultyCache));
 
         // All bindables that should be updated along with the current ruleset + mods.
-        private readonly LockedWeakList<BindableStarDifficulty> trackedBindables = new LockedWeakList<BindableStarDifficulty>();
+        private readonly WeakList<BindableStarDifficulty> trackedBindables = new WeakList<BindableStarDifficulty>();
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapDifficultyCache.cs:34]: Is this change actually required? The issue looks to have been specific to `linkedCancellationSources`.

Comment (smoogipoo) [reply osu.Game/Beatmaps/BeatmapDifficultyCache.cs:34]: Well it isn't, but I'm locking the entire process anyway and thought it was better to remove the double-lock.

Comment (peppy) [reply osu.Game/Beatmaps/BeatmapDifficultyCache.cs:34]: Fair enough, just wanted to make sure I wasn't missing something.

**Review by peppy — APPROVED** [review]
